File: Python/palindromic-path-queries-in-a-tree.py
# Time:  O((n + q) * logn)
# Space: O(nlogn)

# dfs, lca, binary lifting, fenwick tree
class Solution(object):
    def palindromePath(self, n, edges, s, queries):
        """
        :type n: int
        :type edges: List[List[int]]
        :type s: str
        :type queries: List[str]
        :rtype: List[bool]
        """
        class BIT(object):  # 0-indexed.
            def __init__(self, n):
                self.__bit = [0]*(n+1)  # Extra one for dummy node.

            def add(self, i, val):
                i += 1  # Extra one for dummy node.
                while i < len(self.__bit):
                    self.__bit[i] ^= val  # modified
                    i += (i & -i)

            def query(self, i):
                i += 1  # Extra one for dummy node.
                ret = 0
                while i > 0:
                    ret ^= self.__bit[i]  # modified
                    i -= (i & -i)
                return ret


        class TreeInfos(object):  # Time: O(NlogN), Space: O(NlogN), N is the number of nodes
            def __init__(self, adj, cb):  # modified
                N = len(adj)
                L, R, D, P = [0]*N, [0]*N, [0]*N, [[] for _ in xrange(N)]
                idx = -1
                stk = [(1, (0, -1))]
                while stk:
                    step, args = stk.pop()
                    if step == 1:
                        u, p = args
                        cb(u, p)  # added
                        D[u] = 1 if p == -1 else D[p]+1
                        if p != -1:
                            P[u].append(p)
                        i = 0
                        while i < len(P[u]) and i < len(P[P[u][i]]):
                            P[u].append(P[P[u][i]][i])
                            i += 1
                        idx += 1
                        L[u] = idx
                        stk.append((2, (u,)))
                        for i in reversed(xrange(len(adj[u]))):
                            v = adj[u][i]
                            if v == p:
                                continue
                            stk.append((1, (v, u)))
                    elif step == 2:
                        u = args[0]
                        R[u] = idx
                assert(idx == N-1)
                self.L, self.R, self.D, self.P = L, R, D, P

            # Template:
            # https://github.com/kamyu104/FacebookHackerCup-2019/blob/master/Final%20Round/little_boat_on_the_sea.py
            def is_ancestor(self, a, b):  # includes itself
                return self.L[a] <= self.L[b] <= self.R[b] <= self.R[a]

            def lca(self, a, b):
                if self.D[a] > self.D[b]:
                    a, b = b, a
                if self.is_ancestor(a, b):
                    return a
                for i in reversed(xrange(len(self.P[a]))):  # O(logN)
                    if i < len(self.P[a]) and not self.is_ancestor(self.P[a][i], b):
                        a = self.P[a][i]
                return self.P[a][0]


        def callback(u, p):
            prefix[u] = (prefix[p] if p != -1 else 0)^(1<<(ord(s[u])-ord('a')))

        s = list(s)
        adj = [[] for _ in xrange(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
        prefix = [0]*n
        tree_infos = TreeInfos(adj, callback)
        bit = BIT(n+1)
        result = []
        for q in queries:
            args = q.split()
            op = args[0]
            u = int(args[1])
            if op == "update":
                c = args[2]
                diff = (1<<(ord(s[u])-ord('a')))^(1<<(ord(c)-ord('a')))
                if not diff:
                    continue
                s[u] = c
                bit.add(tree_infos.L[u], diff)
                bit.add(tree_infos.R[u]+1, diff)
            else:
                v = int(args[2])
                l = tree_infos.lca(u, v)
                val = (prefix[u]^bit.query(tree_infos.L[u]))^(prefix[v]^bit.query(tree_infos.L[v]))^(1<<(ord(s[l])-ord('a')))
                result.append((val&(val-1)) == 0)
        return result


# Time:  O((n + q) * logn)
# Space: O(nlogn)
# dfs, lca, binary lifting, fenwick tree
class Solution2(object):
    def palindromePath(self, n, edges, s, queries):
        """
        :type n: int
        :type edges: List[List[int]]
        :type s: str
        :type queries: List[str]
        :rtype: List[bool]
        """
        class BIT(object):  # 0-indexed.
            def __init__(self, n):
                self.__bit = [0]*(n+1)  # Extra one for dummy node.

            def add(self, i, val):
                i += 1  # Extra one for dummy node.
                while i < len(self.__bit):
                    self.__bit[i] ^= val  # modified
                    i += (i & -i)

            def query(self, i):
                i += 1  # Extra one for dummy node.
                ret = 0
                while i > 0:
                    ret ^= self.__bit[i]  # modified
                    i -= (i & -i)
                return ret


        class TreeInfos(object):  # Time: O(NlogN), Space: O(NlogN), N is the number of nodes
            def __init__(self, adj):
                N = len(adj)
                L, R, D, P = [0]*N, [0]*N, [0]*N, [[] for _ in xrange(N)]
                idx = -1
                stk = [(1, (0, -1))]
                while stk:
                    step, args = stk.pop()
                    if step == 1:
                        u, p = args
                        D[u] = 1 if p == -1 else D[p]+1
                        if p != -1:
                            P[u].append(p)
                        i = 0
                        while i < len(P[u]) and i < len(P[P[u][i]]):
                            P[u].append(P[P[u][i]][i])
                            i += 1
                        idx += 1
                        L[u] = idx
                        stk.append((2, (u,)))
                        for i in reversed(xrange(len(adj[u]))):
                            v = adj[u][i]
                            if v == p:
                                continue
                            stk.append((1, (v, u)))
                    elif step == 2:
                        u = args[0]
                        R[u] = idx
                assert(idx == N-1)
                self.L, self.R, self.D, self.P = L, R, D, P

            # Template:
            # https://github.com/kamyu104/FacebookHackerCup-2019/blob/master/Final%20Round/little_boat_on_the_sea.py
            def is_ancestor(self, a, b):  # includes itself
                return self.L[a] <= self.L[b] <= self.R[b] <= self.R[a]

            def lca(self, a, b):
                if self.D[a] > self.D[b]:
                    a, b = b, a
                if self.is_ancestor(a, b):
                    return a
                for i in reversed(xrange(len(self.P[a]))):  # O(logN)
                    if i < len(self.P[a]) and not self.is_ancestor(self.P[a][i], b):
                        a = self.P[a][i]
                return self.P[a][0]

        s = list(s)
        adj = [[] for _ in xrange(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
        tree_infos = TreeInfos(adj)
        bit = BIT(n+1)
        for u in xrange(n):
            diff = 1<<(ord(s[u])-ord('a'))
            bit.add(tree_infos.L[u], diff)
            bit.add(tree_infos.R[u]+1, diff)
        result = []
        for q in queries:
            args = q.split()
            op = args[0]
            u = int(args[1])
            if op == "update":
                c = args[2]
                diff = (1<<(ord(s[u])-ord('a')))^(1<<(ord(c)-ord('a')))
                if not diff:
                    continue
                s[u] = c
                bit.add(tree_infos.L[u], diff)
                bit.add(tree_infos.R[u]+1, diff)
            else:
                v = int(args[2])
                l = tree_infos.lca(u, v)
                val = bit.query(tree_infos.L[u])^bit.query(tree_infos.L[v])^(1<<(ord(s[l])-ord('a')))
                result.append(val == 0 or (val&(val-1)) == 0)
        return result

File: C++/palindromic-path-queries-in-a-tree.cpp
// Time:  O((n + q) * logn)
// Space: O(nlogn)

// dfs, lca, binary lifting, fenwick tree
class Solution {
public:
    vector<bool> palindromePath(int n, vector<vector<int>>& edges, string s, vector<string>& queries) {
        vector<int> prefix(n);
        const auto& callback = [&](int u, int p) {
            prefix[u] = (p != -1 ? prefix[p] : 0) ^ ( 1<< (s[u] - 'a'));
        };

        vector<vector<int>> adj(n);
        for (const auto& e : edges) {
            adj[e[0]].emplace_back(e[1]);
            adj[e[1]].emplace_back(e[0]);
        }
        TreeInfos tree_infos(adj, callback);
        BIT bit(n + 1);
        vector<bool> result;
        for (const auto& q : queries) {
            stringstream ss(q);
            string op;
            int u;
            ss >> op >> u;
            if (op == "update") {
                char c;
                ss >> c;
                const auto& diff = (1 << (s[u] - 'a')) ^ (1 << (c - 'a'));
                if (!diff) {
                    continue;
                }
                s[u] = c;
                bit.add(tree_infos.left(u), diff);
                bit.add(tree_infos.right(u) + 1, diff);
            } else {
                int v;
                ss >> v;
                const auto& l = tree_infos.lca(u, v);
                const auto& val = (prefix[u] ^ bit.query(tree_infos.left(u))) ^ (prefix[v] ^ bit.query(tree_infos.left(v))) ^ (1 << (s[l] - 'a'));
                result.emplace_back((val & (val - 1)) == 0);
            }
        }
        return result;
    }

private:
    class BIT {
    public:
        BIT(int n) : bit_(n + 1) {  // 0-indexed
        }
        
        void add(int i, int val) {
            ++i;
            for (; i < size(bit_); i += lower_bit(i)) {
                bit_[i] ^= val;
            }
        }

        int query(int i) const {
            ++i;
            int total = 0;
            for (; i > 0; i -= lower_bit(i)) {
                total ^= bit_[i];
            }
            return total;
        }

    private:
        inline int lower_bit(int i) const {
            return i & -i;
        }
        
        vector<int> bit_;
    };

    class TreeInfos {
    public:
        TreeInfos(const vector<vector<int>>& adj, const auto& cb)  // modified
         : L_(size(adj))
         , R_(size(adj))
         , D_(size(adj))
         , P_(size(adj)) {

             const int N = size(adj);
             int idx = -1;
             vector<tuple<int, int, int>> stk = {{1, 0, -1}};
             while (!empty(stk)) {
                const auto [step, u, p] = stk.back(); stk.pop_back();
                if (step == 1) {
                    cb(u, p);  // added
                    D_[u] = (p == -1) ? 1 : D_[p] + 1;
                    if (p != -1) {
                        P_[u].emplace_back(p);  // ancestors of the node i
                    }
                    for (int i = 0; i < size(P_[u]); ++i) {
                        if (i >= size(P_[P_[u][i]])) {
                            break;
                        }
                        P_[u].emplace_back(P_[P_[u][i]][i]);
                    }
                    L_[u] = ++idx; 
                    stk.emplace_back(2, u, -1);
                    for (int i = size(adj[u]) -1; i >= 0; --i) {
                        const auto& v = adj[u][i];
                        if (v == p) {
                            continue;
                        }
                        stk.emplace_back(1, v, u);
                    }
                } else if (step == 2) {
                    R_[u] = idx;
                }
            }
            assert(idx == N - 1);
        }
        
        bool is_ancestor(int a, int b) const {
            return L_[a] <= L_[b] && R_[b] <= R_[a];
        }

        int lca(int a, int b) const {
            if (D_[a] > D_[b]) {
                swap(a, b);
            }
            if (is_ancestor(a, b)) {
                return a;
            }
            for (int i = size(P_[a]) - 1; i >= 0; --i) {  // O(logN)
                if (i < size(P_[a]) && !is_ancestor(P_[a][i], b)) {
                    a = P_[a][i];
                }
            }
            return P_[a][0];
        }

        int left(int a) const {
            return L_[a];
        }

        int right(int a) const {
            return R_[a];
        }
        
        int depth(int a) const {
            return D_[a];
        }
    
    private:
        vector<int> L_;
        vector<int> R_;
        vector<int> D_;
        vector<vector<int>> P_;
    };
};

// Time:  O((n + q) * logn)
// Space: O(nlogn)
// dfs, lca, binary lifting, fenwick tree
class Solution2 {
public:
    vector<bool> palindromePath(int n, vector<vector<int>>& edges, string s, vector<string>& queries) {
        vector<vector<int>> adj(n);
        for (const auto& e : edges) {
            adj[e[0]].emplace_back(e[1]);
            adj[e[1]].emplace_back(e[0]);
        }
        TreeInfos tree_infos(adj);
        BIT bit(n + 1);
        for (int u = 0; u < n; ++u) {
            const auto& diff = 1 << (s[u] - 'a');
            bit.add(tree_infos.left(u), diff);
            bit.add(tree_infos.right(u) + 1, diff);
        }
        vector<bool> result;
        for (const auto& q : queries) {
            stringstream ss(q);
            string op;
            int u;
            ss >> op >> u;
            if (op == "update") {
                char c;
                ss >> c;
                const auto& diff = (1 << (s[u] - 'a')) ^ (1 << (c - 'a'));
                if (!diff) {
                    continue;
                }
                s[u] = c;
                bit.add(tree_infos.left(u), diff);
                bit.add(tree_infos.right(u) + 1, diff);
            } else {
                int v;
                ss >> v;
                const auto& l = tree_infos.lca(u, v);
                const auto& val = bit.query(tree_infos.left(u)) ^ bit.query(tree_infos.left(v)) ^ (1 << (s[l] - 'a'));
                result.emplace_back(val == 0 || (val & (val - 1)) == 0);
            }
        }
        return result;
    }

private:
    class BIT {
    public:
        BIT(int n) : bit_(n + 1) {  // 0-indexed
        }
        
        void add(int i, int val) {
            ++i;
            for (; i < size(bit_); i += lower_bit(i)) {
                bit_[i] ^= val;
            }
        }

        int query(int i) const {
            ++i;
            int total = 0;
            for (; i > 0; i -= lower_bit(i)) {
                total ^= bit_[i];
            }
            return total;
        }

    private:
        inline int lower_bit(int i) const {
            return i & -i;
        }
        
        vector<int> bit_;
    };

    class TreeInfos {
    public:
        TreeInfos(const vector<vector<int>>& adj)
         : L_(size(adj))
         , R_(size(adj))
         , D_(size(adj))
         , P_(size(adj)) {

             const int N = size(adj);
             int idx = -1;
             vector<tuple<int, int, int>> stk = {{1, 0, -1}};
             while (!empty(stk)) {
                const auto [step, u, p] = stk.back(); stk.pop_back();
                if (step == 1) {
                    D_[u] = (p == -1) ? 1 : D_[p] + 1;
                    if (p != -1) {
                        P_[u].emplace_back(p);  // ancestors of the node i
                    }
                    for (int i = 0; i < size(P_[u]); ++i) {
                        if (i >= size(P_[P_[u][i]])) {
                            break;
                        }
                        P_[u].emplace_back(P_[P_[u][i]][i]);
                    }
                    L_[u] = ++idx; 
                    stk.emplace_back(2, u, -1);
                    for (int i = size(adj[u]) -1; i >= 0; --i) {
                        const auto& v = adj[u][i];
                        if (v == p) {
                            continue;
                        }
                        stk.emplace_back(1, v, u);
                    }
                } else if (step == 2) {
                    R_[u] = idx;
                }
            }
            assert(idx == N - 1);
        }
        
        bool is_ancestor(int a, int b) const {
            return L_[a] <= L_[b] && R_[b] <= R_[a];
        }

        int lca(int a, int b) const {
            if (D_[a] > D_[b]) {
                swap(a, b);
            }
            if (is_ancestor(a, b)) {
                return a;
            }
            for (int i = size(P_[a]) - 1; i >= 0; --i) {  // O(logN)
                if (i < size(P_[a]) && !is_ancestor(P_[a][i], b)) {
                    a = P_[a][i];
                }
            }
            return P_[a][0];
        }

        int left(int a) const {
            return L_[a];
        }

        int right(int a) const {
            return R_[a];
        }
        
        int depth(int a) const {
            return D_[a];
        }
    
    private:
        vector<int> L_;
        vector<int> R_;
        vector<int> D_;
        vector<vector<int>> P_;
    };
};
